src/main/java/org/openpnp/gui/GcodeDriverConsole.java

```diff
@@ -0,0 +1,200 @@
+        lblCmdLine = new JLabel("Command line:");
+        gcodeConsole.add(lblCmdLine, "2, 4");
+        
+        cmdLineTextField = new JTextField();              Logger.debug("keyCode = " + e.getKeyCode() + ", historyCursor = " + historyCursor);
+
+        gcodeConsole.add(cmdLineTextField,"4, 4");
+        
+        cmdLineTextField.addKeyListener(new KeyAdapter() {
+            @Override
+            public void keyPressed(KeyEvent e) {
+            
+              switch (e.getKeyCode()) {
+                case 0:  return;
+                
+                case 38: // arrow key UP, older history
+                  if (historyCursor < historyLen - 1) {
+                    historyCursor++;
+                  }
+                  cmdLineTextField.setText(history[historyCursor]);
+                  break;
+                  
+                case 40:  // arrow key DOWN, more recent history
+                  if (historyCursor > 0) {
+                    historyCursor--;
+                  }
+                  cmdLineTextField.setText(history[historyCursor]);
+                  break;
+                  
+                default:
+                  if (e.getKeyChar() == '\n') {
+                    sendGcodeConCmd();
+                  }
+                  else {
+                      super.keyTyped(e);
+                  }
+                  break;
+              }
+            }
+          }
+        );
+
+        sendGcodeConCmdBtn = new JButton(sendGcodeConCmdAction);
+        gcodeConsole.add(sendGcodeConCmdBtn, "6, 4");
+
+    }
+    @Override
+    public void createBindings() {
+        IntegerConverter intConverter = new IntegerConverter();
+        DoubleConverter doubleConverter =
+                new DoubleConverter(Configuration.get().getLengthDisplayFormat());
+        
+		}
+    private JTextArea textAreaConsole;
+    private JLabel lblCmdLine;
+    private JTextField cmdLineTextField;
+    private JButton sendGcodeConCmdBtn;
+    
+    private String history[];
+    private int historyLen;
+    private int historyMaxLen;
+    private int historyCursor;
+        
+    private String getHistoryAtIdx() {
+      return history[historyCursor];
+    }
+    
+    private void sendGcodeConCmd() {
+    
+      String cmd = cmdLineTextField.getText();
+      int moveLen = 0;
+      
+      // are we repeating the same command? Store it in history only if different than last one
+      if (!cmd.equals(history[0])) {
+        if (historyLen < historyMaxLen) {
+          moveLen = historyLen++;
+        } else {
+          moveLen = historyMaxLen - 1;
+        }
+        // shift history one place to the past. Copy is done in-place.
+        System.arraycopy(history, 0, history, 1, moveLen);
+        // store current command at 0
+        history[0] = cmd;
+      }
+      // reset historyCursor to the present
+      historyCursor = 0;
+      
+      // most controllers prefer commands to be in upper case.
+      cmd = cmd.toUpperCase();
```

Comment:
I don't see any reason to force upper case. I don't know of any controllers that require it and it could restrict someone from using a command that requires mixed case.